### ISSUE TITLE
Could not debug. error: more than one device/emulator

### ISSUE THREAD

**Issue by Ruzo** [Microsoft/vscode-react-native#324]
## Actual Behavior
1. Using one device through tcpip
2. Builds successfully
3. Stops debugging with error: "Could not debug. error: more than one device/emulator"

## Expected Behavior
1. Using one device through tcpip
2. Builds successfully
3. Debugging stays active

## Software versions
- React-Native VS Code extension version: 0.21
- VSCode version: 1.61
- OS platform and version: Windows 10 Enterprise 1607
- React Native version: 0.35

## Outputs (Include if relevant)
- Output of the Debug Console (View -> Toggle Debug Console): 

BUILD SUCCESSFUL

Total time: 21.414 secs
Starting the app on 192.168.1.168:5555 (C:\Users\Richard\AppData\Local\Android\sdk/platform-tools/adb -s 192.168.1.168:5555 shell am start -n com.manager/.MainActivity)...
Starting: Intent { cmp=com.manager/.MainActivity }

[vscode-react-native] Finished executing: react-native.cmd run-android
[Error] "Could not debug. error: more than one device/emulator"

- Output of the React-Native output channel (View -> Toggle Output -> Select React-Native in ListBox): 
- Output of the Developer Tools console (Help -> Toggle Developer Tools -> Select Console tab):

**Comment by Ruzo** [comment]
Works fine with Genymotion simulator.

**Comment by MSLaguana** [comment]
Could you please open a command line and run `adb devices` and show the response you get?

**Comment by MSLaguana** [comment]
Could you please clarify the scenario where it doesn't work? When you have a network attached device, but not with a network attached emulator or with a physically attached device?

**Comment by Ruzo** [comment]
Don't know if it's related but, from time to time after launching VScode, I see an error message in the output: "Failed to setup the stub launcher for the debugger". It doesn't show up every time.

**Comment by MSLaguana** [comment]
I can replicate your error but only if `adb devices` shows up multiple targets (e.g. multiple emulator instances).

You should be able to work around this by explicitly specifying the device you want to connect to in `launch.json`: add a `"target": "192.168.8.8:5555",` property to the "launch android" entry, where the value is whatever shows up for your device under `adb devices`.

**Comment by Ruzo** [comment]
Still the same error with the "target" property inserted in "Debug Android" task in launch.json. Actually, the phone does not connect at all to the server when I make that change.

Thanks for trying to help but I have a feeling that it's my own private little bug with adb. For some reason, only part of the Debug console output could be pasted to my description but I now can see an error at the beginning of the build which is:
```
Running C:\Users\Richard\AppData\Local\Android\sdk/platform-tools/adb -s 192.168.1.168:5555 reverse tcp:8081 tcp:8081
error: more than one device/emulator
Could not run adb reverse: Command failed: C:\Users\Richard\AppData\Local\Android\sdk/platform-tools/adb -s 192.168.1.168:5555 reverse tcp:8081 tcp:8081
```
trying the "adb -s 192.168.1.168:5555 reverse tcp:8081 tcp:8081" command in a Command Prompt gives the same "error: more than one device/emulator"

**Comment by gmaliandi** [comment]
Exact same problem here. Running `adb -s 192.168.0.100:5555 reverse tcp:8081 tcp:8081` always gives `"error: more than one device/emulator"`

**Comment by Ruzo** [comment]
@gmaliandi, still the same for me. I thought it had to do with Genymotion because of VirtualBox but that wasn't it. Does it have to do with Samsung's driver since I'm using a Note 4? I'm waiting for my pixel to know for sure.

**Comment by DFelten** [comment]
Same error here if I try to run on a device without cable:

```
cd C:\Users\Daniel\AppData\Local\Android\sdk\platform-tools
adb tcpip 5555 && adb connect 192.168.1.129:5555
react-native start
react-native run-android
```

Could not get BatchedBridge, make sure your bundle is packaged correctly.

`[Error : 0104]  Failed to run the application in Android: error: more than one device/emulator`

```
adb devices
List of devices attached
192.168.1.129:5555      device
```

**Comment by vladimir-kotikov** [comment]
@DFelten, if you want to debug over wi-fi, make sure that you have specified your debug host IP and port in app's Developer menu -> Dev settings -> Debug server host & port

**Comment by vladimir-kotikov** [comment]
@gmaliandi, @Ruzo, were you guys trying to debug app that was connected through network, rather than cable?

**Comment by DFelten** [comment]
Thanks, after defining the IP and my port it's working :)

**Comment by Ruzo** [comment]
@vladimir-kotikov yes, through network with host IP and port entered in Dev settings. It worked before, not sure what happened. Did not try it since though.

**Comment by Dashue** [comment]
Same issue even when configured in "Debug server host & port".
Wouldn't consider this issue to be closed

**Comment by careergod** [comment]
same problem here , 

error: more than one device/emulator
Could not run adb reverse: Command failed: /Users/Macbookpro2/Library/Android/sdk/platform-tools/adb -s 10.61.147.228:8081 reverse tcp:8081 tcp:8081

**Comment by iTOYS** [comment]
Hi @careergod 
Please see [this issue](https://github.com/facebook/react-native/issues/17379) in react-native repo.
Seems like there is some problems with android adb

**Comment by Nermin-K** [comment]
Are there any other solutions?! I have tried everything above and nothing helps. I have stuck there for 4 days.

**Comment by nicotroia** [comment]
If there is an emulator stuck on "connecting" when you do `adb devices` then that means there is a program running within the adb port range. (`emulator-5562` for me running on 5562 was stuck on "connecting").

I discovered that the Native Instruments Access program which I installed, runs `NTKDaemon` on my computer (on port 5562). In order to build and start my app on emulator I have to force close NTKDaemon. On mac os you can open `Activity Monitor` to kill it, or Control panel on windows.

Hope this helps someone

**Comment by omar-bakhsh** [comment]
if you have this Error  
`Couldn't start project on Android: Error running adb: more than one device/emulator`
this means you have 2 android devices connected **( your mobile by USB ), COUNTS**  **dISCONNECT** and Retry Again

**Comment by Sanan4li** [comment]
anyone solved this issue? I have same error even there is only on device which is connected over wifi. I have same wifi network connected on the machine as well as device.
It works fine on USB cable but I want to get rid of the cable.

**Comment by SounD120** [comment]
Hi @Sanan4li. I could debug android application remotely over wi-fi using the following steps:
1. Install application on device and run it.
2. Set debug server host and port in `Dev Settings -> Debug server host & port for device` on device pointing to IP-address of my dev machine on which packager has been started.
3. Start packager on dev machine.
4. Open application folder in VS Code and run `Attach to debugger` configuration.
6. Enable remote JS debugging in application on device.

Could you please let us know, if it is that what are you looking for?

**Comment by danfmaia** [comment]
> Still the same error with the "target" property inserted in "Debug Android" task in launch.json. Actually, the phone does not connect at all to the server when I make that change.
> 
> Thanks for trying to help but I have a feeling that it's my own private little bug with adb. For some reason, only part of the Debug console output could be pasted to my description but I now can see an error at the beginning of the build which is:
> 
> ```
> Running C:\Users\Richard\AppData\Local\Android\sdk/platform-tools/adb -s 192.168.1.168:5555 reverse tcp:8081 tcp:8081
> error: more than one device/emulator
> Could not run adb reverse: Command failed: C:\Users\Richard\AppData\Local\Android\sdk/platform-tools/adb -s 192.168.1.168:5555 reverse tcp:8081 tcp:8081
> ```
> 
> trying the "adb -s 192.168.1.168:5555 reverse tcp:8081 tcp:8081" command in a Command Prompt gives the same "error: more than one device/emulator"

It is not your private little error. I'm getting the same error here. My `adb devices` also outputs a single connected device, when connected via tcpip:
```
List of devices attached
192.168.0.71:5555       device
```

**Comment by Ruzo** [comment]
> It is **not** your private little bug. I'm getting the same error here. My `adb devices` also outputs a single connected device, when connected via tcpip:
> 
> ```
> List of devices attached
> 192.168.0.71:5555       device
> ```

Sorry @danfmaia. I never found a solution and using Flutter now. Maybe you could try some of the solutions posted above.

**Comment by RedMickey** [comment]
Hi @danfmaia and @Ruzo, have you tried to attach to a running application following [steps in this comment](https://github.com/Microsoft/vscode-react-native/issues/324#issuecomment-576156266)?

**Comment by RedMickey** [comment]
We've released React Native Tools v1.5.1 with improvements for debugging on remote Android devices. Now launch scenarios shouldn't fail due to the `error: more than one device/emulator` error.
